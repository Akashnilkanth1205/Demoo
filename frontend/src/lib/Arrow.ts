/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"

import { IStyler } from "autogen/proto"

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

interface TableDimensions {
  headerRows: number
  headerColumns: number
  dataRows: number
  dataColumns: number
  rows: number
  columns: number
}

type TableCellType = "blank" | "index" | "columns" | "data"

interface TableCell {
  type: TableCellType
  id?: string
  classNames: string
  content: string
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: Arrow
}

export class Arrow {
  private table: Table

  private schema: Schema

  private index: string[][]

  private columns: string[][]

  private data: any[][]

  private styler?: Styler

  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
    this.table = Table.from(dataBuffer)
    this.schema = this.getSchema()
    this.index = this.getIndex()
    this.columns = this.getColumns()
    this.data = this.getData()
    this.styler = {
      uuid: styler?.uuid || undefined,
      caption: styler?.caption || undefined,
      styles: styler?.styles || undefined,
      displayValues: styler?.displayValues
        ? new Arrow(styler.displayValues)
        : undefined,
    }
  }

  private getSchema(): Schema {
    const schema = this.table.schema.metadata.get("pandas")
    if (schema == null) {
      throw new Error("Table schema is missing.")
    }
    return JSON.parse(schema)
  }

  private static getColumnData(column: Column): any[] {
    return range(0, column.length).map(rowIndex => column.get(rowIndex))
  }

  private getIndex(): string[][] {
    return this.schema.index_columns.map(field => {
      const isRangeIndex = typeof field === "object" && field.kind === "range"
      if (isRangeIndex) {
        const { start, stop, step } = field as RangeIndex
        return range(start, stop, step).map(String)
      }
      const column = this.table.getColumn(field as string)
      return Arrow.getColumnData(column).map(String)
    })
  }

  private getColumns(): string[][] {
    const isMultiIndex = this.schema.column_indexes.length > 1
    return unzip(
      this.schema.columns
        .map(column => column.field_name)
        .filter(fieldName => !this.schema.index_columns.includes(fieldName))
        .map(fieldName =>
          isMultiIndex
            ? JSON.parse(
                fieldName
                  .replace(/\(/g, "[")
                  .replace(/\)/g, "]")
                  .replace(/'/g, '"')
              )
            : [fieldName]
        )
    )
  }

  private getData(): any[][] {
    const rows = this.table.length
    const columns = this.columns.length > 0 ? this.columns[0].length : 0

    return range(0, rows).map(rowIndex =>
      range(0, columns).map(columnIndex =>
        this.table
          .getColumnAt(columnIndex)
          ?.get(rowIndex)
          ?.toString()
      )
    )
  }

  get tableId(): string | undefined {
    return this.styler?.uuid && `T_${this.styler.uuid}`
  }

  get tableStyles(): string | undefined {
    return this.styler?.styles
  }

  get caption(): string | undefined {
    return this.styler?.caption
  }

  get dimensions(): TableDimensions {
    const [headerColumns, dataRowsCheck] = this.index.length
      ? [this.index.length, this.index[0].length]
      : [0, 0]

    const [headerRows, dataColumnsCheck] = this.columns.length
      ? [this.columns.length, this.columns[0].length]
      : [0, 0]

    const [dataRows, dataColumns] = this.data.length
      ? [this.data.length, this.data[0].length]
      : // If there is no data, default to the number of header columns.
        [0, dataColumnsCheck]

    if (
      (dataRows !== 0 && dataRows !== dataRowsCheck) ||
      (dataColumns !== 0 && dataColumns !== dataColumnsCheck)
    ) {
      throw new Error(
        "Table dimensions don't align: " +
          `rows(${dataRows} != ${dataRowsCheck}) OR ` +
          `cols(${dataColumns} != ${dataColumnsCheck})`
      )
    }

    const rows = headerRows + dataRows
    const columns = headerColumns + dataColumns

    return {
      headerRows,
      headerColumns,
      dataRows,
      dataColumns,
      rows,
      columns,
    }
  }

  public getCell(rowIndex: number, columnIndex: number): TableCell {
    const { headerRows, headerColumns, rows, columns } = this.dimensions

    if (rowIndex < 0 || rowIndex >= rows) {
      throw new Error("Row index is out of range.")
    }
    if (columnIndex < 0 || columnIndex >= columns) {
      throw new Error("Column index is out of range.")
    }

    const isBlankCell = rowIndex < headerRows && columnIndex < headerColumns
    const isIndexCell = rowIndex >= headerRows && columnIndex < headerColumns
    const isColumnsCell = rowIndex < headerRows && columnIndex >= headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push(`level${rowIndex}`)
      }
      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    }

    if (isIndexCell) {
      const dataRowIndex = rowIndex - headerRows

      const uuid = this.styler?.uuid
      const classNames = [
        `row_heading`,
        `level${columnIndex}`,
        `row${dataRowIndex}`,
      ]

      return {
        type: "index",
        id: uuid
          ? `T_${uuid}level${columnIndex}_row${dataRowIndex}`
          : undefined,
        classNames: classNames.join(" "),
        // Table index is stored as is (in the column format).
        content: this.index[columnIndex][dataRowIndex],
      }
    }

    if (isColumnsCell) {
      const dataColumnIndex = columnIndex - headerColumns

      const classNames = [
        `col_heading`,
        `level${rowIndex}`,
        `col${dataColumnIndex}`,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.columns[rowIndex][dataColumnIndex],
      }
    }

    const dataRowIndex = rowIndex - headerRows
    const dataColumnIndex = columnIndex - headerColumns

    const uuid = this.styler?.uuid
    const classNames = ["data", `row${dataRowIndex}`, `col${dataColumnIndex}`]

    const content = this.styler?.displayValues
      ? this.styler.displayValues.getCell(rowIndex, columnIndex).content
      : this.data[dataRowIndex][dataColumnIndex]

    return {
      type: "data",
      id: uuid
        ? `T_${uuid}row${dataRowIndex}_col${dataColumnIndex}`
        : undefined,
      classNames: classNames.join(" "),
      content,
    }
  }
}
