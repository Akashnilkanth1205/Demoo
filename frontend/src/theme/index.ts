/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LightTheme, DarkTheme } from "baseui"
import { lightBaseUITheme, darkBaseUITheme } from "./baseui"
import base from "./baseTheme"
import light from "./lightTheme"
import dark from "./darkTheme"
import sidebar from "./sidebarTheme"
import { Theme } from "./types"

export * from "./baseui"
export * from "./types"
export * from "./utils"
export const lightTheme: Theme = light
export const darkTheme: Theme = dark
export const sidebarTheme: Theme = sidebar

export const AvailableTheme = {
  lightTheme: {
    name: "Light",
    emotion: light,
    baseweb: LightTheme,
    basewebTheme: lightBaseUITheme,
  },
  darkTheme: {
    name: "Dark",
    emotion: dark,
    baseweb: DarkTheme,
    basewebTheme: darkBaseUITheme,
  },
  customTheme: {
    // TODO: base, main or nothing?
    name: "Custom",
    emotion: base,
    baseweb: LightTheme,
    basewebTheme: lightBaseUITheme,
  },
}
