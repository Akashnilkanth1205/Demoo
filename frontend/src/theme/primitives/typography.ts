export const fonts = {
  sansSerif: "IBM Plex Sans, sans-serif",
  mono: "IBM Plex Mono, monospace",
}

export const fontSizes = {
  twoXS: "0.375rem",
  xs: "0.5rem",
  sm: "0.75rem",
  smDefault: "0.8rem",
  md: "1rem",
  lg: "1.25rem",
  xl: "1.5rem",
  twoXL: "1.75rem",
  threeXL: "2.25rem",
  fourXL: "3.125rem",
}

export const fontWeights = {
  normal: 400,
  medium: 500,
  bold: 700,
}

export const lineHeights = {
  normal: "normal",
  none: "1",
  tight: "1.25",
  base: "1.6",
}

export const letterSpacings = {
  tighter: "-0.05em",
  tight: "-0.025em",
  normal: "0",
  wide: "0.025em",
  wider: "0.05em",
  widest: "0.1em",
}
