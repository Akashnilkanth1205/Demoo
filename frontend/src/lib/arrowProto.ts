/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"
import { Data, Index, IndexType, Quiver } from "src/lib/Quiver"

export interface DataFrame {
  index: Index
  columns: string[][]
  data: Data
  styler?: Styler
}

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface IndexData {
  data: string[]
  type: IndexType
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

export enum IndexTypes {
  UnicodeIndex = "unicode",
  RangeIndex = "range",
  CategoricalIndex = "categorical",
  MultiIndex = "multiIndex",
  IntervalIndex = "interval[int64]",
  DatetimeIndex = "datetime",
  TimedeltaIndex = "time",
  PeriodIndex = "period[Q-DEC]",
  Int64Index = "int64",
  UInt64Index = "uint64",
  Float64Index = "float64",
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: any
}

export function parseTable(table: Table): DataFrame {
  const schema = getSchema(table)
  const index = {
    data: getIndex(table, schema),
    type: getIndexType(schema),
  }
  const columns = getColumns(schema)
  const data = {
    data: getData(table, columns),
    type: getColumnType(schema),
  }
  return {
    index,
    columns,
    data,
  }
}

function getSchema(table: Table): Schema {
  const schema = table.schema.metadata.get("pandas")
  if (schema == null) {
    throw new Error("Table schema is missing.")
  }
  return JSON.parse(schema)
}

function getIndex(table: Table, schema: Schema): string[][] {
  return schema.index_columns.map(field => {
    const isRangeIndex = typeof field === "object" && field.kind === "range"
    if (isRangeIndex) {
      const { start, stop, step } = field as RangeIndex
      return range(start, stop, step).map(String)
    }
    const column = table.getColumn(field as string)
    return getColumnData(column).map(String)
  })
}

function getIndexType(schema: Schema): IndexType[] {
  return schema.index_columns.map(indexName => {
    if (isRangeIndex(indexName)) {
      return {
        name: IndexTypes.RangeIndex,
        meta: indexName,
      }
    }

    // Get index column from columns schema.
    const indexColumn = schema.columns.find(
      column => column.field_name === indexName
    )

    // PeriodIndex and IntervalIndex values are kept in `numpy_type` property,
    // the rest in `pandas_type`.
    return {
      name:
        indexColumn?.pandas_type === "object"
          ? indexColumn?.numpy_type
          : indexColumn?.pandas_type,
      meta: indexColumn?.metadata,
    }
  })
}

function getColumns(schema: Schema): string[][] {
  const isMultiIndex = schema.column_indexes.length > 1
  return unzip(
    schema.columns
      .map(column => column.field_name)
      .filter(fieldName => !schema.index_columns.includes(fieldName))
      .map(fieldName =>
        isMultiIndex
          ? JSON.parse(
              fieldName
                .replace(/\(/g, "[")
                .replace(/\)/g, "]")
                .replace(/'/g, '"')
            )
          : [fieldName]
      )
  )
}

function getColumnType(schema: Schema): any {
  return schema.columns
    .filter(column => !schema.index_columns.includes(column.field_name))
    .map(column => column.pandas_type)
}

function getData(table: Table, columns: string[][]): string[][] {
  const rows = table.length
  const cols = columns.length > 0 ? columns[0].length : 0

  return range(0, rows).map(rowIndex =>
    range(0, cols).map(columnIndex =>
      table.getColumnAt(columnIndex)?.get(rowIndex)
    )
  )
}

function getColumnData(column: Column): string[] {
  return range(0, column.length).map(rowIndex => column.get(rowIndex))
}

function isRangeIndex(field: string | RangeIndex): boolean {
  return typeof field === "object" && field.kind === "range"
}

export function concatTables(q1: Quiver, q2: Quiver): DataFrame {
  // Special case if df1 is empty.
  if (q1.data.data.length === 0) {
    return {
      index: q2.index,
      data: q2.data,
      columns: q2.columns,
    }
  }

  // Always return df1 columns
  const { columns } = q1
  const index = concatIndices(q1.index, q2.index)
  const data = concatData(q1.data, q2.data)
  return { index, data, columns }
}

function concatIndices(i1: Index, i2: Index): Index {
  // Special case if i1 is empty.
  if (i1.data.length === 0) {
    return i2
  }

  // Otherwise, make sure the types match.
  if (!sameIndexTypes(i1.type, i2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(i1.type)} with ${JSON.stringify(
        i2.type
      )}.`
    )
  }

  return i1.data.reduce(
    (newIndex: Index, indexData: string[], index: number) => {
      const concatenatedIndex = concatIndex(
        indexData,
        i2.data[index],
        i1.type[index]
      )
      newIndex.data.push(concatenatedIndex.data)
      newIndex.type.push(concatenatedIndex.type)
      return newIndex
    },
    { data: [], type: [] }
  )
}

function concatIndex(i1: string[], i2: string[], type: IndexType): IndexData {
  // Special case for RangeIndex.
  if (type.name === IndexTypes.RangeIndex) {
    return concatRangeIndex(i1, i2, type)
  }
  return concatAnyIndex(i1, i2, type)
}

function concatRangeIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  let newStop = type.meta.stop
  return i2.reduce(
    (newIndex: IndexData) => {
      newIndex.data.push(newStop)
      newStop += type.meta.step
      newIndex.type.meta.stop = newStop
      return newIndex
    },
    { data: i1, type }
  )
}

function concatAnyIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  const concatenatedIndex = i1.concat(i2)

  // Special case for CategoricalIndex:
  // `num_categories` must be increased.
  if (type.name === IndexTypes.CategoricalIndex) {
    type.meta.num_categories += i2.length
  }

  return { data: concatenatedIndex, type }
}

function concatData(d1: Data, d2: Data): Data {
  // Make sure data types match.
  if (!sameColumnTypes(d1.type, d2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(d1.type)} with ${JSON.stringify(
        d2.type
      )}.`
    )
  }

  const numberOfColumns = d1.type.length
  return {
    data: d1.data.concat(
      d2.data.map((data: string[]) => data.slice(0, numberOfColumns))
    ),
    type: d1.type,
  }
}

function sameIndexTypes(t1: IndexType[], t2: IndexType[]): boolean {
  // Make sure both indices have same dimensions.
  if (t1.length !== t2.length) {
    return false
  }

  return t1.every(
    (firstTypeValue: IndexType, index: number) =>
      firstTypeValue.name === t2[index].name
  )
}

function sameColumnTypes(firstType: string[], secondType: string[]): boolean {
  return firstType.every(
    (firstTypeValue: string, index: number) =>
      firstTypeValue === secondType[index]
  )
}
