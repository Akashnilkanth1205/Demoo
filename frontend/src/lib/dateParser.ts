import parser, { Parser, Eat, Tokenizer } from "remark-parse"
import { DateTimeHandler } from "lib/DateTime"
import { Datetime } from "autogen/proto"
function dates() {
  //@ts-ignore
  var Parser = this.Parser
  var tokenizers = Parser.prototype.inlineTokenizers
  var methods = Parser.prototype.inlineMethods

  tokenizers.datetime = tokenizeDatetimes as Tokenizer
  // Run it just before `text`.
  methods.splice(methods.indexOf("text"), 0, "datetime")
}

tokenizeDatetimes.notInLink = true
tokenizeDatetimes.locator = locateDate

function locateDate(value: string, fromIndex: number) {
  return value.indexOf("{@", fromIndex)
}

function tokenizeDatetimes(eat: Eat, value: string, silent: boolean) {
  var match = /{@datetime: (.*)}(.*$)/.exec(value)
  if (match) {
    if (silent) {
      return true
    }

    const date = Datetime.fromObject(JSON.parse(match[1]))
    console.log(JSON.parse(match[1]), date)
    // Reconstruct markdown
    const markdownParts = [
      { type: "inlineCode", value: DateTimeHandler.dateToString(date) },
    ]
    if (match.index) {
      markdownParts.unshift({
        type: "text",
        value: value.substring(0, match.index),
      })
    }
    if (match[2]) {
      markdownParts.push({ type: "text", value: match[2] })
    }

    return eat(value)({
      type: "root",
      children: markdownParts,
    })
  }
}

export default dates
