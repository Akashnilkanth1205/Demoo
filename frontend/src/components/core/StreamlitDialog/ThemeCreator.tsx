import React, { ReactElement } from "react"
import { toHex } from "color2k"
import humanizeString from "humanize-string"
import { Check } from "@emotion-icons/material-outlined"
import { CustomThemeConfig } from "autogen/proto"
import PageLayoutContext from "components/core/PageLayoutContext"
import Button, { Kind } from "components/shared/Button"
import UISelectbox from "components/shared/Dropdown"
import Icon from "components/shared/Icon"
import { createTheme, ThemeConfig, toThemeInput } from "theme"
import {
  StyledButtonContainer,
  StyledHeader,
  StyledSmall,
  StyledThemeColorPicker,
  StyledThemeCreator,
  StyledThemeCreatorWrapper,
  StyledThemeDesc,
} from "./styled-components"

interface ThemeOptionBuilder {
  desc: string
  title: string
  component: any
  options?: any[]
  getValue: (value: string, config: ThemeOptionBuilder) => any
}

const valueToColor = (value: string, _config: ThemeOptionBuilder): string =>
  toHex(value).toUpperCase()

const displayFontOption = (
  font: CustomThemeConfig.FontFamily | string
): string =>
  // @ts-ignore
  humanizeString(CustomThemeConfig.FontFamily[font])

const themeBuilder: Record<string, ThemeOptionBuilder> = {
  primaryColor: {
    desc:
      "Used to style primary interface elements. Displayed most frequently across your app's screens and components.",
    title: "Primary color",
    component: StyledThemeColorPicker,
    getValue: valueToColor,
  },
  secondaryColor: {
    desc:
      "(Optional) Used to style secondary interface elements. It provides ways to accent and distinguish your app.",
    title: "Secondary color",
    component: StyledThemeColorPicker,
    getValue: valueToColor,
  },
  backgroundColor: {
    desc: "Background color for the main container.",
    title: "Background color",
    component: StyledThemeColorPicker,
    getValue: valueToColor,
  },
  secondaryBackgroundColor: {
    desc:
      "Used as the background for most widgets. Examples of widgets with this background are st.sidebar, st.text_input, st.date_input.",
    title: "Secondary background color",
    component: StyledThemeColorPicker,
    getValue: valueToColor,
  },
  textColor: {
    desc: "Font color for the page",
    title: "Text color",
    component: StyledThemeColorPicker,
    getValue: valueToColor,
  },
  font: {
    desc:
      "Font family (serif | sans serif | monospace) for the page. Will not impact the code areas.",
    title: "Font family",
    options: Object.keys(CustomThemeConfig.FontFamily).map(font =>
      humanizeString(font)
    ),
    getValue: (value: string, config: ThemeOptionBuilder): number =>
      (config.options &&
        config.options.findIndex(
          (font: string) => font === displayFontOption(value)
        )) ||
      0,
    component: UISelectbox,
  },
}

const ThemeCreator = (): ReactElement => {
  const [copied, updateCopied] = React.useState(false)
  const [isOpen, openCreator] = React.useState(false)
  const themeCreator = React.useRef<HTMLDivElement>(null)
  const { activeTheme, addThemes, setTheme } = React.useContext(
    PageLayoutContext
  )

  const themeInput = toThemeInput(activeTheme.emotion)

  const updateTheme = (customTheme: ThemeConfig): void => {
    addThemes([customTheme])
    setTheme(customTheme)
  }

  const onThemeOptionChange = (key: string, newVal: string): void => {
    const customTheme = createTheme({
      ...themeInput,
      [key]: newVal,
      name: "Custom Theme",
    })
    updateTheme(customTheme)
    updateCopied(false)
  }

  const config = `[theme]
primaryColor="${themeInput.primaryColor}"
secondaryColor="${themeInput.secondaryColor}"
backgroundColor="${themeInput.backgroundColor}"
secondaryBackgroundColor="${themeInput.secondaryBackgroundColor}"
textColor="${themeInput.textColor}"
font="${displayFontOption(
    themeInput.font || CustomThemeConfig.FontFamily.SANS_SERIF
  ).toLowerCase()}"
`

  const toggleCreatorUI = (): void => {
    openCreator(true)
  }

  React.useEffect(() => {
    if (isOpen && themeCreator.current) {
      themeCreator.current.scrollIntoView(true)
    }
  }, [isOpen])

  const copyConfig = (): void => {
    navigator.clipboard.writeText(config)
    updateCopied(true)
  }

  const renderThemeOption = (
    themeOption: string,
    value: string
  ): ReactElement | null => {
    const themeOptionConfig = themeBuilder[themeOption]
    if (themeOptionConfig === undefined) return null

    const isColor = themeOptionConfig.component === StyledThemeColorPicker
    // Props that vary based on component type
    const variableProps = {
      options: themeOptionConfig.options || undefined,
      showValue: isColor,
      value: themeOptionConfig.getValue(value, themeOptionConfig),
    }
    return (
      <React.Fragment key={themeOption}>
        <themeOptionConfig.component
          disabled={false}
          label={themeOptionConfig.title}
          onChange={(newVal: string) =>
            onThemeOptionChange(themeOption, newVal)
          }
          {...variableProps}
        />
        <StyledThemeDesc>{themeOptionConfig.desc}</StyledThemeDesc>
      </React.Fragment>
    )
  }
  return (
    <StyledThemeCreatorWrapper ref={themeCreator}>
      {isOpen ? (
        <>
          <StyledHeader>Edit active theme</StyledHeader>
          <p>
            Change settings below to see live changes to your active theme. To
            discard changes and reload the original themes, refresh the page.
          </p>
          <StyledThemeCreator>
            {Object.entries(themeInput).map(([themeOption, value]) =>
              renderThemeOption(themeOption, value as string)
            )}
          </StyledThemeCreator>

          <StyledButtonContainer>
            <Button onClick={copyConfig} kind={Kind.PRIMARY}>
              {copied ? (
                <>
                  {"Copied to clipboard "}
                  <Icon
                    content={Check}
                    size="lg"
                    color={activeTheme.emotion.colors.success}
                  />
                </>
              ) : (
                "Copy theme to clipboard"
              )}
            </Button>
            <StyledSmall>
              {copied
                ? "Paste copied theme to config.toml to save theme"
                : "Copy the above settings in TOML format"}
            </StyledSmall>
          </StyledButtonContainer>
        </>
      ) : (
        <Button onClick={toggleCreatorUI} kind={Kind.LINK}>
          Edit active theme
        </Button>
      )}
    </StyledThemeCreatorWrapper>
  )
}

export default ThemeCreator
