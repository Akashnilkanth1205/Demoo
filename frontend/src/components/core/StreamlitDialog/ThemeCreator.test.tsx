/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { CustomThemeConfig } from "autogen/proto"
import { mount, shallow } from "lib/test_util"
import ColorPicker from "components/shared/ColorPicker"
import UISelectbox from "components/shared/Dropdown"
import { baseTheme } from "theme"
import { fonts } from "theme/primitives/typography"
import ThemeCreator, { Props } from "./ThemeCreator"

const mockSetTheme = jest.fn()
const mockAddThemes = jest.fn()

const getProps = (extend?: Partial<Props>): Props => ({
  hasCustomTheme: false,
  ...extend,
})

Object.assign(navigator, {
  clipboard: {
    writeText: jest.fn(),
  },
})

window.HTMLElement.prototype.scrollIntoView = jest.fn()

describe("Renders ThemeCreator", () => {
  beforeEach(() =>
    jest.spyOn(React, "useContext").mockImplementation(() => ({
      activeTheme: baseTheme,
      setTheme: mockSetTheme,
      availableThemes: [],
      addThemes: mockAddThemes,
    }))
  )

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("renders closed theme creator without custom theme", () => {
    const props = getProps()
    const wrapper = shallow(<ThemeCreator {...props} />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find("Button").prop("children")).toBe(
      "Create a new Custom Theme"
    )
  })

  it("Renders closed theme creator with custom theme", () => {
    const props = getProps({ hasCustomTheme: true })
    const wrapper = shallow(<ThemeCreator {...props} />)
    expect(wrapper.find("Button").prop("children")).toBe(
      "Edit Existing Custom Theme"
    )
  })

  it("Renders opened theme creator", () => {
    const props = getProps()
    const wrapper = shallow(<ThemeCreator {...props} />)
    wrapper.find("Button").simulate("click")
    expect(wrapper).toMatchSnapshot()
  })
})

describe("Opened ThemeCreator", () => {
  beforeEach(() =>
    jest.spyOn(React, "useContext").mockImplementation(() => ({
      activeTheme: baseTheme,
      setTheme: mockSetTheme,
      availableThemes: [],
      addThemes: mockAddThemes,
    }))
  )

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("should update theme on color change", () => {
    const props = getProps()
    const wrapper = mount(<ThemeCreator {...props} />)
    wrapper.find("Button").simulate("click")

    const colorpicker = wrapper.find(ColorPicker)
    expect(colorpicker).toHaveLength(5)

    colorpicker.at(0).prop("onChange")("pink")
    expect(mockAddThemes).toHaveBeenCalled()
    expect(mockAddThemes.mock.calls[1][0][0].emotion.colors.primary).toBe(
      "pink"
    )

    expect(mockSetTheme).toHaveBeenCalled()
    expect(mockSetTheme.mock.calls[1][0].emotion.colors.primary).toBe("pink")
  })

  it("should update theme on font change", () => {
    const props = getProps()
    const wrapper = mount(<ThemeCreator {...props} />)
    wrapper.find("Button").simulate("click")
    const selectbox = wrapper.find(UISelectbox)
    const { options } = selectbox.props()

    expect(options).toHaveLength(
      Object.keys(CustomThemeConfig.FontFamily).length
    )

    selectbox.prop("onChange")(2)
    expect(mockAddThemes).toHaveBeenCalled()
    expect(
      mockAddThemes.mock.calls[1][0][0].emotion.genericFonts.bodyFont
    ).toBe(fonts.monospace)

    expect(mockSetTheme).toHaveBeenCalled()
    expect(mockSetTheme.mock.calls[1][0].emotion.genericFonts.bodyFont).toBe(
      fonts.monospace
    )
  })

  it("should copy to clipboard", () => {
    const { colors } = baseTheme.emotion
    const props = getProps()
    const wrapper = mount(<ThemeCreator {...props} />)
    wrapper.find("Button").simulate("click")
    const copyBtn = wrapper.find("Button")

    expect(copyBtn.prop("children")).toBe("Copy Theme to Clipboard")
    copyBtn.simulate("click")
    // TODO: Karrie; fix font expected result
    // font="sans serif"
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(`[theme]
primaryColor="${colors.primary}"
secondaryColor="${colors.secondary}"
backgroundColor="${colors.bgColor}"
secondaryBackgroundColor="${colors.secondaryBg}"
textColor="${colors.bodyText}"
font="0"
`)
  })
})
