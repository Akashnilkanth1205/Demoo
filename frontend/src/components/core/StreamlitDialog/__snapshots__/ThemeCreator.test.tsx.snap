// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ThemeCreator Renders opened theme creator 1`] = `
<StyledThemeCreatorWrapper>
  <StyledHeader>
    Edit active theme
  </StyledHeader>
  <p>
    Changes exist for the duration of a session. To discard changes and recover the original themes, refresh the page.
  </p>
  <StyledThemeCreator>
    <StyledThemeColorPicker
      disabled={false}
      label="Primary color"
      onChange={[Function]}
      showValue={true}
      value="#F63366"
    />
    <StyledThemeDesc>
      Used to style primary interface elements. Displayed most frequently across your app's screens and components.
    </StyledThemeDesc>
    <StyledThemeColorPicker
      disabled={false}
      label="Secondary color"
      onChange={[Function]}
      showValue={true}
      value="#A3A8B8"
    />
    <StyledThemeDesc>
      (Optional) Used to style secondary interface elements. It provides ways to accent and distinguish your app.
    </StyledThemeDesc>
    <StyledThemeColorPicker
      disabled={false}
      label="Background color"
      onChange={[Function]}
      showValue={true}
      value="#FFFFFF"
    />
    <StyledThemeDesc>
      Background color for the main container.
    </StyledThemeDesc>
    <StyledThemeColorPicker
      disabled={false}
      label="Secondary background color"
      onChange={[Function]}
      showValue={true}
      value="#F0F2F6"
    />
    <StyledThemeDesc>
      Used as the background for most widgets. Examples of widgets with this background are st.sidebar, st.text_input, st.date_input.
    </StyledThemeDesc>
    <StyledThemeColorPicker
      disabled={false}
      label="Text color"
      onChange={[Function]}
      showValue={true}
      value="#262730"
    />
    <StyledThemeDesc>
      Font color for the page.
    </StyledThemeDesc>
    <Selectbox
      disabled={false}
      label="Font family"
      onChange={[Function]}
      options={
        Array [
          "Sans serif",
          "Serif",
          "Monospace",
        ]
      }
      showValue={false}
      value={0}
    />
    <StyledThemeDesc>
      Font family (serif | sans serif | monospace) for the page. Will not impact the code areas.
    </StyledThemeDesc>
  </StyledThemeCreator>
  <StyledPasteInstructions>
    <StyledSmall>
      To save as a Theme, paste settings in the 'theme' section in your
    </StyledSmall>
    <code>
      config.toml
    </code>
    <StyledSmall>
      file.
    </StyledSmall>
  </StyledPasteInstructions>
  <StyledButtonContainer>
    <Button
      kind="primary"
      onClick={[Function]}
    >
      Copy theme to clipboard
    </Button>
  </StyledButtonContainer>
</StyledThemeCreatorWrapper>
`;

exports[`Renders ThemeCreator renders closed theme creator without custom theme 1`] = `
<StyledThemeCreatorWrapper>
  <Button
    kind="link"
    onClick={[Function]}
  >
    Edit active theme
  </Button>
</StyledThemeCreatorWrapper>
`;
